### PR TITLE
[wip] Addition of `TextTester` class.

### LINKED ISSUE #411: `TextTester` class based in `MetricTester` class to robustly test `text` metrics
## 🚀 Feature

The current `text` package uses various functions to test the `text` metrics. The `MetricTester` class provides a much more robust way to test metrics on all fronts such as multi-device setup etc.

Hence, we can create a new class called `TextTester` based on `MerticTester` that provides us with the necessary tools to robustly test `text` metrics. 

### Motivation

The addition of a new class is required as the inputs to text based metrics are different where they take an `str` or a `Sequence[str]` and not `Tensor` type unlike metrics from other domains.

### Pitch

An implementation of the following functions and class methods that support inputs as text (`str` or `Sequence[str]`) which can be used to test any `text` metric. 
```
def _class_test();

def _functional_test():


class TextTester:
    def run_functional_metric_test():

    def run_class_metric_test():

    @staticmethod
    def run_precision_test_cpu():

    @staticmethod
    def run_precision_test_gpu():

    @staticmethod
    def run_differentiability_test():
```
This will also follow the conversion of the currently existing metrics to `TextTester` based testing.

### Alternatives

We could just use the current function based approach.

### PR DESCRIPTION
**Pull request by karthikrangasai**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?

Fixes #411 .

## To Do:
[x] Add TextTester class.
[ ] Update testing for BLEU Score
[ ] Update testing for ROUGE Metric
[ ] Update testing for WER
[ ] Update testing for BertScore

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/text/bleu.py [hunk 1]
```diff
@@ -55,8 +55,22 @@ class BLEUScore(Metric):
...
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Optional[Callable] = None,
+    ):
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
```
Comment (SkafteNicki) [inline torchmetrics/text/bleu.py:71]: remember to add(copy from some other metric) description in docstring

File: tests/text/test_wer.py [hunk 2]
```diff
@@ -14,55 +19,57 @@
...
+
+
+@pytest.mark.skipif(not _JIWER_AVAILABLE, reason="test requires jiwer")
+class TestWER(TextTester):
+    @pytest.mark.parametrize("ddp", [False, True])
+    @pytest.mark.parametrize("dist_sync_on_step", [False, True])
+    @pytest.mark.parametrize(
+        ["preds", "targets"],
+        [
+            pytest.param(BATCHES["preds"], BATCHES["targets"]),
```
Comment (SkafteNicki) [inline tests/text/test_wer.py:36]: could we add a second case, just to make sure that the metric is not hand engineered to work on this single case. Here is a example from jiwer:

```
ground_truth = ["i like monthy python", "what do you mean, african or european swallow"]
hypothesis = ["i like", "python", "what you mean" , "or swallow"]

ground_truth = ["hello world", "i like monthy python"]
hypothesis = ["hello duck", "i like python"]

```

File: tests/text/test_wer.py [hunk 3]
```diff
@@ -14,55 +19,57 @@
...
-    assert pl_output == jiwer_output
+BATCHES = {"preds": [[PREDICTION1], [PREDICTION2]], "targets": [[REFERENCE1], [REFERENCE2]]}
+
+
+def _compute_wer_metric_jiwer(prediction: Union[str, List[str]], reference: Union[str, List[str]]):
+    return compute_measures(reference, prediction)["wer"]
+
+
+@pytest.mark.skipif(not _JIWER_AVAILABLE, reason="test requires jiwer")
+class TestWER(TextTester):
```
Comment (SkafteNicki) [inline tests/text/test_wer.py:30]: insert it here to parametrize all tests with preds and target one time
```suggestion
@pytest.mark.parametrize(
        ["preds", "targets"],
        [
            pytest.param(BATCHES["preds"], BATCHES["targets"]),
        ],
    )
class TestWER(TextTester):
```

File: tests/text/test_rouge.py [hunk 4]
```diff
@@ -163,18 +74,54 @@ def test_rouge_metric_class(pl_rouge_metric_key, use_stemmer):
...
-    assert torch.allclose(pl_output[pl_rouge_metric_key], rs_result)
+class TestBLEUScore(TextTester):
+    @pytest.mark.parametrize("ddp", [False, True])
+    @pytest.mark.parametrize("dist_sync_on_step", [False, True])
+    @pytest.mark.parametrize(
+        ["preds", "targets"],
+        [
+            pytest.param(BATCHES["preds"], BATCHES["targets"]),
+        ],
+    )
```
Comment (SkafteNicki) [inline tests/text/test_rouge.py:85]: again, move in front of class

File: tests/text/helpers.py [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+try:
+    set_start_method("spawn")
+except RuntimeError:
+    pass
+
+
+@unique
+class INPUT_ORDER(Enum):
+    PREDS_FIRST = 1
+    TARGETS_FIRST = 2
```
Comment (SkafteNicki) [inline tests/text/helpers.py:38]: so some of the metrics have the `preds, target` order and others have `target, preds`?

File: tests/text/helpers.py [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+    if isinstance(pl_result, Tensor):
+        assert np.allclose(pl_result.detach().cpu().numpy(), sk_result, atol=atol, equal_nan=True)
+    # multi output compare
+    elif isinstance(pl_result, Sequence):
+        for pl_res, sk_res in zip(pl_result, sk_result):
+            _assert_allclose(pl_res, sk_res, atol=atol)
+    elif isinstance(pl_result, Dict):
+        if key is None:
+            raise KeyError("Provide Key for Dict based metric results.")
+        assert np.allclose(pl_result[key].detach().cpu().numpy(), sk_result, atol=atol, equal_nan=True)
```
Comment (SkafteNicki) [inline tests/text/helpers.py:57]: lets just add this part (and the `key` argument) to the original `_assert_allclose` and then you can just import it here

File: tests/text/helpers.py [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+
+def _assert_tensor(pl_result: Any, key: str = None):
+    """Utility function for recursively checking that some input only consists of torch tensors."""
+    if isinstance(pl_result, Sequence):
+        for plr in pl_result:
+            _assert_tensor(plr)
+    elif isinstance(pl_result, Dict):
+        if key is None:
+            raise KeyError("Provide Key for Dict based metric results.")
+        assert isinstance(pl_result[key], Tensor)
```
Comment (SkafteNicki) [inline tests/text/helpers.py:70]: again, just add to base original and import

File: tests/text/helpers.py [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+def _assert_requires_grad(metric: Metric, pl_result: Any, key: str = None):
+    """Utility function for recursively asserting that metric output is consistent with the `is_differentiable`
+    attribute."""
+    if isinstance(pl_result, Sequence):
+        for plr in pl_result:
+            _assert_requires_grad(metric, plr, key=key)
+    elif isinstance(pl_result, Dict):
+        if key is None:
+            raise KeyError("Provide Key for Dict based metric results.")
+        assert metric.is_differentiable == pl_result[key].requires_grad
```
Comment (SkafteNicki) [inline tests/text/helpers.py:84]: also here

File: tests/text/helpers.py [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+    kwargs_update = {
+        k: (v[0].half() if v.is_floating_point() else v[0]).to(device) if isinstance(v, Tensor) else v
+        for k, v in kwargs_update.items()
+    }
+    metric_module = metric_module.to(device)
+    _assert_tensor(metric_module(y_hat, y, **kwargs_update))
+    _assert_tensor(metric_functional(y_hat, y, **kwargs_update))
+
+
+class TextTester:
```
Comment (SkafteNicki) [inline tests/text/helpers.py:300]: what about we do inherence here? then we do not need to have a copy of `setup_class` and `teardown_class`
```suggestion
class TextTester(MetricTester):
```

**Review by SkafteNicki — COMMENTED** [review]
okay, a mix of comments. Looking great so far and will really improve the testing of our text metrics

File: tests/text/helpers.py (hunk 5 above)
Comment (karthikrangasai) [reply tests/text/helpers.py:38]: Yes. This is case with the packages being tested against and the same order is used in our package's implementation as well.

File: tests/text/helpers.py (hunk 9 above)
Comment (karthikrangasai) [reply tests/text/helpers.py:300]: I thought it would cause typing errors, but that didn't happen. So I sub-classed `TextTester` from `MetricTester`.

**Comment by karthikrangasai** [comment]
Updating the implementation of `ROUGE` by adding a state helped get rid of some failed tests. But the class based implementation tests still fail when `ddp=True`.
Can I get insights on this issue ? It would be helpful.

File: tests/helpers/testers.py [hunk 10]
```diff
@@ -57,7 +57,7 @@ def setup_ddp(rank, world_size):
         torch.distributed.init_process_group("gloo", rank=rank, world_size=world_size)
 
 
-def _assert_allclose(pl_result: Any, sk_result: Any, atol: float = 1e-8):
+def _assert_allclose(pl_result: Any, sk_result: Any, atol: float = 1e-8, key: str = None):
```
Comment (SkafteNicki) [inline tests/helpers/testers.py:60]:
```suggestion
def _assert_allclose(pl_result: Any, sk_result: Any, atol: float = 1e-8, key: Optional[str] = None):
```

File: tests/helpers/testers.py [hunk 11]
```diff
@@ -66,25 +66,37 @@ def _assert_allclose(pl_result: Any, sk_result: Any, atol: float = 1e-8):
...
+    elif isinstance(pl_result, Dict):
+        if key is None:
+            raise KeyError("Provide Key for Dict based metric results.")
+        assert np.allclose(pl_result[key].detach().cpu().numpy(), sk_result, atol=atol, equal_nan=True)
     else:
         raise ValueError("Unknown format for comparison")
 
 
-def _assert_tensor(pl_result: Any):
+def _assert_tensor(pl_result: Any, key: str = None):
```
Comment (SkafteNicki) [inline tests/helpers/testers.py:77]:
```suggestion
def _assert_tensor(pl_result: Any, key: Optional[str] = None):
```

File: tests/helpers/testers.py [hunk 12]
```diff
@@ -66,25 +66,37 @@ def _assert_allclose(pl_result: Any, sk_result: Any, atol: float = 1e-8):
...
+    elif isinstance(pl_result, Dict):
+        if key is None:
+            raise KeyError("Provide Key for Dict based metric results.")
+        assert isinstance(pl_result[key], Tensor)
     else:
         assert isinstance(pl_result, Tensor)
 
 
-def _assert_requires_grad(metric: Metric, pl_result: Any):
+def _assert_requires_grad(metric: Metric, pl_result: Any, key: str = None):
```
Comment (SkafteNicki) [inline tests/helpers/testers.py:90]:
```suggestion
def _assert_requires_grad(metric: Metric, pl_result: Any, Optional[key]: str = None):
```

**Review by SkafteNicki — COMMENTED** [review]
going to take a look at the failing test

File: tests/helpers/testers.py [hunk 13]
```diff
@@ -66,25 +66,37 @@ def _assert_allclose(pl_result: Any, sk_result: Any, atol: float = 1e-8):
...
+    elif isinstance(pl_result, Dict):
+        if key is None:
+            raise KeyError("Provide Key for Dict based metric results.")
+        assert isinstance(pl_result[key], Tensor)
     else:
         assert isinstance(pl_result, Tensor)
 
 
-def _assert_requires_grad(metric: Metric, pl_result: Any):
+def _assert_requires_grad(metric: Metric, pl_result: Any, Optional[key]: str = None):
```
Comment (SkafteNicki) [inline tests/helpers/testers.py:90]:
```suggestion
def _assert_requires_grad(metric: Metric, pl_result: Any, key: Optional[str] = None):
```
